File: April-18/10-04-18/java/03-Odd-Even-Subarrays.java

import java.util.*;

class TestClass {
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        int ar[] = new int[n+1];
        ar[0] = 0;
        for(int i = 1; i <= n; i++)
        {
            ar[i] = input.nextInt();
            if(ar[i] % 2 == 0)
                ar[i] = 1;
            else
                ar[i] = -1;
        }
        
        HashMap<Integer,Integer> mp = new HashMap<Integer,Integer>();
        
        mp.put(0 , 1);
        long ans = 0;
        for(int i = 1; i <= n; i++)
        {
            ar[i] += ar[i-1];
            int val = 0;
            if(mp.containsKey(ar[i]))
                val = mp.get(ar[i]);
            
            ans += val;
            mp.put(ar[i] , (val+1));
        }
        System.out.println(ans);
    
        
    }
}

File: April-18/10-04-18/python/03-Odd-Even-Subarrays.py
n = int(input())
data = [int(x) for x in input().split()]

ar = []
ar.append(0)
for i in range(0,n):
    if(data[i] % 2 == 0):
        ar.append(1)
    else:
        ar.append(-1)

mp = {}
mp[0] = 1
ans = 0
for i in range(1,n+1):
    ar[i] += ar[i-1]
    val = 0
    try:
        val = mp[ar[i]]
    except:
        mp[ar[i]] = 0
    ans += val
    mp[ar[i]] += 1
print(ans)
    
